File: LeetCode/1136. Parallel Courses/solution.py
from collections import deque

class Solution:
    def minimumSemesters(self, N: int, relations: List[List[int]]) -> int:
        depend = [[] for _ in range(N+1)]
        income = [0] * (N+1)
        # print(depend)
        for re in relations:
            income[re[0]] += 1
            
            depend[re[1]] += [re[0]]
            
         
        q = deque()
        for i in range(1,N+1,1):
            if income[i] == 0:
                q.append(i)
        
        cc, level = 0, 0
        while len(q) > 0:
            n = len(q)
            cc += n
            for _ in range(n):
                front = q.popleft()
                for dpd in depend[front]:
                    income[dpd] -= 1
                    if income[dpd] == 0:
                        q.append(dpd)
            level += 1
        if cc != N:
            return -1
        return level

File: LeetCode/1136. Parallel Courses/solution.cpp
class Solution {
public:
    int minimumSemesters(int n, vector<vector<int>>& relations) {
        vector<vector<int>> graph(n+1);
        vector<int> indegree(n+1);
        for (auto relation : relations){
            graph[relation[0]].push_back(relation[1]);
            indegree[relation[1]]++;
        }
        queue<int> q;
        int step = 0, cnt = 0;
        for (int i = 1; i <= n; i++)
            if (indegree[i] == 0)
                q.push(i);
        while (!q.empty()){
            int len = q.size();
            for (int j = 0; j < len; j++){
                auto cur = q.front(); q.pop();
                cnt++;
                for (auto nxt : graph[cur]){
                    indegree[nxt]--;
                    if (indegree[nxt] == 0)
                        q.push(nxt);
                }
            }
            step++;
        }
        if (cnt < n)
            return -1;
        return step;
    }
};